### ISSUE TITLE
Write a proptest to check Amount formatting.

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#2954]
If you can write that proptest properly I'd be thrilled. I tried doing this (as a fuzztest, not as a proptest) but I got constantly gummed up by the `width` parameter which behaves weirdly because of the denomination. I'm not even certain how it *should* behave -- does the denomination's width count against the width of the output?

This should probably be promoted to its own issue. I really would like this fuzztest to exist because it would document exactly how we believe Amount should be formatted.

_Originally posted by @apoelstra in https://github.com/rust-bitcoin/rust-bitcoin/issues/2604#issuecomment-2204029408_

**Comment by shinghim** [comment]
Has there been any discussion on which prop testing library we should bring in, if any at all?

**Comment by apoelstra** [comment]
No. I had assumed we'd use `proptest` but I haven't looked into it at all.

**Comment by shinghim** [comment]
Gotcha, I had assumed the same but just wanted to double check. I'll read up on it and put something together

**Comment by Kixunil** [comment]
Crap, I swear I wrote a response here that didn't get posted. I have experience with quickcheck which is pretty easy to use but I'm not sure if it's better than other alternatives. From a quick look `proptest` looks good. Especially having a stable `Arbitrary` trait is nice and we could provide it for our types publicly.

**Comment by yancyribbens** [comment]
> Especially having a stable Arbitrary trait is nice and we could provide it for our types publicly.

Arbitrary requires rustc `1.63.0`, so we'd need to update the MSRV first before adding stable Arbitrary I think.

**Comment by Kixunil** [comment]
IIRC we no longer have policy that all deps must have certain MSRV. Anyway, `proptest` doesn't depend on `arbitrary`. (This is actually unfortunate as it'd be better to just have one trait that works for everything.)

Also MSRV 1.63 was already suggested and there were no objections so far and we have at least one damn good reason to bump it.

**Comment by yancyribbens** [comment]
> proptest doesn't depend on arbitrary

Personally I think it would be advantageous to create proptests that use Arbitrary, although it's true that it's not a req.

> Also MSRV 1.63 was already suggested and there were no objections so far and we have at least one damn good reason to bump it.

Yeah I saw that discussed.  Agree that it's a good reason to bump to 1.63.  I'm not sure what kind of lift that is considering all the crate smashing going on.

**Comment by Kixunil** [comment]
> I think it would be advantageous to create proptests that use Arbitrary

Why exactly?

> 'm not sure what kind of lift that is considering all the crate smashing going on.

Should be just a simple update of all MSRV references

**Comment by yancyribbens** [comment]
> Why exactly?

For the same reason that it would be nice to use Arbitrary with fuzz testing?  Fuzz tests do not require Arbitrary either but it makes it more convenient.

**Comment by apoelstra** [comment]
Should we investigate https://crates.io/crates/arbtest instead?

Alternately, does `proptest` have an `Arbitrary`-like trait that we could use in place of `Arbitrary`?

We definitely ought to be able to share code between our arbitrary stuff and our proptest stuff since they're so closely related. There is [this old blog post which talks about using bolero to do this](https://model-checking.github.io/kani-verifier-blog/2022/10/27/using-kani-with-the-bolero-property-testing-framework.html) though `bolero` *also* doesn't seem to depend on `Arbitrary`.

**Comment by shinghim** [comment]
> Should we investigate https://crates.io/crates/arbtest instead?

I think this would be worth looking into since everyone here seems to agree that it would be nice to use the same [`Arbitrary`](https://crates.io/crates/arbitrary) trait for fuzzing and proptesting. IMO the minimal implementation is nice, and if it can get the job done + provide the flexibility of using `Arbitrary`, I think we should use it

> Alternately, does proptest have an Arbitrary-like trait that we could use in place of Arbitrary?

Looks like `proptest` has [their own version of `Arbitrary`](https://altsysrq.github.io/rustdoc/proptest/latest/proptest/arbitrary/index.html), which uses a `Strategy` that can add some customization to how the arbitrary values are created.

**Comment by apoelstra** [comment]
Ok, the proptest `Arbitrary` trait actually looks way more powerful and useful than the one from the `arbitrary` crate. I vote that we use the proptest one and forget about `arbitrary`.

**Comment by yancyribbens** [comment]
> Ok, the proptest Arbitrary trait actually looks way more powerful and useful than the one from the arbitrary crate. I vote that we use the proptest one and forget about arbitrary.

Will this work with honggfuzz?

**Comment by apoelstra** [comment]
@yancyribbens yes -- the honggfuzz tests will just manually call methods from the traits, so they'll "work with" anything at all.

**Comment by yancyribbens** [comment]
I've not used proptest before so I thought I'd give it a shot

I wrote a proptest for BnB coin-selection: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/60

> the honggfuzz tests will just manually call methods from the traits, so they'll "work with" anything at all.

Now that I know a little bit about proptest, I still don't see a clean way to do this.  From what I can tell, proptest uses it's own source of entropy, and so does honggfuzz.  Are you proposing to just ignore the entropy from honggfuzz in the fuzz tests and use arbitrary structured data created by proptest?  Surely I'm missing something..

**Comment by apoelstra** [comment]
https://docs.rs/proptest/latest/proptest/test_runner/enum.RngAlgorithm.html talks about fuzzer entropy

**Comment by shinghim** [comment]
IIUC, it looks like we could ignore `honggfuzz` in favor of the `proptest::arbitrary` implementation

The current fuzz tests look like:
```rust
fn main() {
    loop {
        fuzz!(|data| {
            do_test(data);
        });
    }
}
```

If we create an arbitrary value with `proptest::arbitrary`, we wouldn't be using the `data` variable provided by `honggfuzz`, so we wouldn't need the `fuzz!` macro:
```rust
fn main() {
    loop {
        let some_instance = SomeType::arbitrary()
            .new_tree(&mut proptest::test_runner::TestRunner::default())
            .unwrap()
            .current();
        do_test(some_instance);
    }
}
```

**Comment by apoelstra** [comment]
What proptest does and what a fuzzer does are completely different. One cannot replace the other. But I would like to reuse code to the extent possible between them.

**Comment by yancyribbens** [comment]
> If we create an arbitrary value with proptest::arbitrary, we wouldn't be using the data variable provided by honggfuzz, so we wouldn't need the fuzz! macro:

This is what I meant by saying that the I don't see a clean way to combine honggfuzz and proptest arbitrary.  Ignoring the entropy created by honggfuzz destroys the usefulness of honggfuzz imo.  The Arbitrary crate https://docs.rs/arbitrary/latest/arbitrary/ produces structured data using the entropy produced by honggfuzz which is better (imo).

**Comment by shinghim** [comment]
It sounds like the robust tests via `proptest` are more valuable than code reuse between the proptests and fuzz tests, so I'll move forward with that

**Comment by Kixunil** [comment]
Isn't hongfuzz more suited for testing *parsers* rather than producing arbitrary values? Anyway, we could probably have a macro for defining `arbitrary::Arbitrary` and `proptest::arbitrary::Arbitrary` at the same time.

**Comment by apoelstra** [comment]
Why would we need `arbitrary::Arbitrary` if we already had `proptest::arbitrary::Arbitrary`? Why not just use the latter. It seems strictly better.

It can still be used to "manually" construct arbitrary objects in a fuzzer-driven way, even if we don't want to use the direct fuzzer support that `proptest` has which I linked above.

And yes, fuzzers are much better at testing parsers than at producing arbitrary values -- unless you write "generative fuzztests" where you "parse" arbitrary bytestrings into arbitrary values using an `Arbitrary`-like trait.

**Comment by Kixunil** [comment]
Crap, I completely forgot you've inked it then went to look for it didn't find it so I assumed I misremembered it.

**Comment by yancyribbens** [comment]
> Why would we need arbitrary::Arbitrary if we already had proptest::arbitrary::Arbitrary? Why not just use the latter. It seems strictly better.

I would add however that:

- `proptest::arbitrary::Arbitrary` can't be used since it's MSRV is `1.65` (although they document the MSRV is `1.64`, so I non-current version is requred. `Arbitrary` however could be used with it's latest release version since they match the now current `1.63` MSRV of rust-bitcoin.
- `proptest::arbitrary::Arbitrary` can only be used `#[cfg_attr(test,` currently while `Arbitrary` can be used as a optional without `test` config.

I agree `proptest::arbitrary::Arbitrary`  looks strictly better _if_ we could use the current version and without cfg_attr test.

**Comment by Kixunil** [comment]
> * `proptest::arbitrary::Arbitrary` can't be used since it's MSRV is `1.65`

There's nothing stopping us writing a test that only runs on that version. Or even 1.80 if we want.



> * `proptest::arbitrary::Arbitrary` can only be used `#[cfg_attr(test,` currently while `Arbitrary` can be used as a optional without `test` config.

I don't understand this. Which attribute do you want to make conditional on `test`?

**Comment by apoelstra** [comment]
>There's nothing stopping us writing a test that only runs on that version. Or even 1.80 if we want.

True, though it will make my life much harder since we will then have dependencies in the lockfile that I can't actually build with our MSRV. It would require some pretty ugly hacks for me to make that work with my local CI, like pruning the lockfile conditioned on compiler version.

Having said that, I suspect proptest 1.3 or 1.4 will work with 1.63. Certainly 1.0 will work since it long predates 1.63. So we may just have to restrict ourselves to the featureset of an old version.

As for the `cfg(test)` stuff, that sounds like a recommendation to avoid polluting our API. But I think we want to pollute our API because we want downstream users to also be able to generate arbitrary objects for testing purposes.

**Comment by yancyribbens** [comment]
> Having said that, I suspect proptest 1.3 or 1.4 will work with 1.63. Certainly 1.0 will work since it long predates 1.63. So we may just have to restrict ourselves to the featureset of an old version.

I had to go back to `1.2` before I found a version that worked with `1.63`

> As for the cfg(test) stuff, that sounds like a recommendation to avoid polluting our API.

I found it to be a requirement.  It's documented here: https://proptest-rs.github.io/proptest/proptest-derive/getting-started.html under using derive.  There's a PR linked to fix it from long ago that's still open.

The cfg(test) thing makes it especially difficult for downstream crates to depend on I would think.  Maybe it's worth sticking with `Arbitrary` for now and `proptest::arbitrary::Arbitrary` for use only with proptest.

**Comment by Kixunil** [comment]
> since we will then have dependencies in the lockfile that I can't actually build with our MSRV

I believe this is only a problem if they use some new cargo features. But OK, let's try to avoid that. The trait will be available in 1.0 anyway.

@yancyribbens that link speaks about derive. We can impl manually (or write a decl macro).

Also I think it's OK to have an optional feature that breaks under non-test usage (maybe it's even good thing?) although it might make our CI annoying.

**Comment by yancyribbens** [comment]
> We can impl manually (or write a decl macro).

@Kixunil I think this is overkill for providing arbitrary values to fuzz with.  To implement `proptest::arbitrary::Arbitrary`, you must also implement `Parameters` and `Strategy`.  `Strategy` not only defines how to generate random values, but also how to shrink values to simpler forms.  To implement strategy, one must implement Tree and Value, to facilitate shrinking.  However shrinking is really only useful for protests and not fuzzing.

In otherwords, there's a lot of boilerplate that needs to be created for each type, a lot of which isn't useful for fuzz testing.

Take a look at the sample file here to see what I'm talking about: https://github.com/proptest-rs/proptest/blob/master/proptest/src/arbitrary/sample.rs

**Comment by Kixunil** [comment]
@yancyribbens we actually want to do both. And we want to provide the impl for downstream consumers too. So it doesn't matter that it's not useful for fuzzing if we need it for something else anyway.

**Comment by apoelstra** [comment]
>However shrinking is really only useful for protests and not fuzzing.

Definitely not true.

**Comment by yancyribbens** [comment]
> Definitely not true.

Ok I stand corrected.  But what I did find is that honggfuzz does not implement shrinking.

**Comment by apoelstra** [comment]
Yes. I have been meaning to move away from honggfuzz to libfuzzer for this reason but it's not a super high priority.

**Comment by yancyribbens** [comment]
> @yancyribbens we actually want to do both. And we want to provide the impl for downstream consumers too. So it doesn't matter that it's not useful for fuzzing if we need it for something else anyway.

Thanks for the response, I'll look into this further.  As @apoelstra pointed out, it could be useful for fuzzing (which I didn't realize).  Currently as a downstream consumer I'm not using this functionality, but it is interesting to learn more about.

**Comment by Kixunil** [comment]
> Currently as a downstream consumer I'm not using this functionality, but it is interesting to learn more about.

I, as a downstream user, do use property testing and I had to do some super horrible Rust black magic hacks to implement `Arbitrary`. I'd definitely appreciate removing that mess by depending on `bitcoin`'s impls even if it means moving from `quickcheck` to `proptest`.

**Comment by yancyribbens** [comment]
> Yes. I have been meaning to move away from honggfuzz to libfuzzer for this reason but it's not a super high priority.

I had actually implemented libfuzzer initially then switched to honggfuzz because I read libfuzzer is no longer actively being worked on https://llvm.org/docs/LibFuzzer.html and also to maintain parity with the rust-bitcoin ecosystem.

**Comment by yancyribbens** [comment]
What I came up with for Arbitrary proptest Amount:

```
impl Arbitrary for Amount {
    type Parameters = <u64 as Arbitrary>::Parameters;
    type Strategy = strategy::Map<<u64 as Arbitrary>::Strategy, fn(u64) -> Amount>;

    fn arbitrary_with(p: Self::Parameters) -> Self::Strategy {
        Strategy ::prop_map(any_with::<u64> (p), |a| Amount { 0: a })
    }
}
```

**Comment by Kixunil** [comment]
@yancyribbens That looks easy but it'd be nice to check if LLVM can devirtualize the calls since this is performance sensitive code.

**Comment by yancyribbens** [comment]
> That looks easy but it'd be nice to check if LLVM can devirtualize the calls since this is performance sensitive code.

I'd say that this is performance sensitive code during run-time in terms of fuzzing, since run-time performance hits will stackup as fuzzing repeats over days/weeks/months, however this isn't the case with proptesting.  If we can say that polymorphism happens at compile time, then I think this is a non-issue, which I think it ought to since all the types can be known to the compiler at compile time AFAICT.  That is, if this can be shown to use compile-time polymorphism instead of run-time polymorphism, then I don't think there is a considerable performance hit.  For reference, I'm basing _my opinions_ somewhat on the information presented here: https://stackoverflow.com/questions/28621980/what-are-the-actual-runtime-performance-costs-of-dynamic-dispatch.

excerpt for link ref:
```
devirtualization: sometimes the compiler can resolve the virtual table look-up at compile time
(usually, because it knows the concrete type of the object); if so, it can therefore use a regular
function call rather than an indirect one, and optimize away the indirection.
```

One way to know for sure I suppose would be to observe the performance of fuzzing using the Arbitrary trait and without.

**Comment by Kixunil** [comment]
> however this isn't the case with proptesting

Why wouldn't it? Prop testing is very similar to fuzzing.

I know about devirtualization but it'd be best to check if it's in fact happening.

**Comment by yancyribbens** [comment]
> Why wouldn't it? Prop testing is very similar to fuzzing.

I just meant to say that the number of iterations a proptest runs is negligible in comparison to a fuzztest.  For example, I have a fuzz test that's been running now for a few weeks, where as a proptest finishes in a few seconds.  This leads me to conclude that performance changes won't be noticeable for proptests in the way that it would be for fuzz tests.  Of course this is just my anecdotal experience.

**Comment by yancyribbens** [comment]
> I know about devirtualization but it'd be best to check if it's in fact happening.

Sounds good.  How is a good way to do that?

**Comment by Kixunil** [comment]
> For example, I have a fuzz test that's been running now for a few weeks, where as a proptest finishes in a few seconds.

IIUC this is just a matter of configuration not a fundamental property (LOL, pun unintended) of property tests. Still IIUC we want to use the impls for fuzzing too?

> Sounds good. How is a good way to do that?

Check the generated assembly. If it messes around with the function pointer then it's not optimized.

**Comment by yancyribbens** [comment]
> Still IIUC we want to use the impls for fuzzing too?

Yes, if you recall, I have this PR open https://github.com/rust-bitcoin/rust-bitcoin/pull/3015, however, this issue was opened, and it was decided to not use Arbitrary but instead the Arbitrary from proptest.  Does that make sense?

> Check the generated assembly. If it messes around with the function pointer then it's not optimized.

OK, although I haven't much experience with that, although in principle it makes sense.

**Comment by yancyribbens** [comment]
```rust
impl Arbitrary for Amount {
    type Parameters = <u64 as Arbitrary>::Parameters;
    type Strategy = strategy::Map<<u64 as Arbitrary>::Strategy, fn(u64) -> Amount>;

    fn arbitrary_with(p: Self::Parameters) -> Self::Strategy {
        any_with::<u64> (p).prop_map(|a| Amount { 0: a })
    }
}
```

The generated assembly:
```asm
.section ".text.<bitcoin_units::amount::Amount as proptest::arbitrary::traits::Arbitrary>::arbitrary_with","ax",@progbits
        .globl  <bitcoin_units::amount::Amount as proptest::arbitrary::traits::Arbitrary>::arbitrary_with
        .p2align        4, 0x90
        .type   <bitcoin_units::amount::Amount as proptest::arbitrary::traits::Arbitrary>::arbitrary_with,@function
<bitcoin_units::amount::Amount as proptest::arbitrary::traits::Arbitrary>::arbitrary_with:
        .cfi_startproc
        push rax
        .cfi_def_cfa_offset 16
        mov rax, qword ptr [rip + __rust_no_alloc_shim_is_unstable@GOTPCREL]
        movzx eax, byte ptr [rax]
        mov edi, 24
        mov esi, 8
        call qword ptr [rip + __rust_alloc@GOTPCREL]
        test rax, rax
        je .LBB42_2
        mov qword ptr [rax], 1
        mov qword ptr [rax + 8], 1
        lea rcx, [rip + core::ops::function::FnOnce::call_once]
        mov qword ptr [rax + 16], rcx
        pop rcx
        .cfi_def_cfa_offset 8
        ret
.LBB42_2:
        .cfi_def_cfa_offset 16
        mov edi, 8
        mov esi, 24
        call qword ptr [rip + alloc::alloc::handle_alloc_error@GOTPCREL]
```

note, this is build with release profile:
```
Finished `release` profile [optimized] target(s) in 0.15s
```

**Comment by Kixunil** [comment]
Oh, that's not devirtualized but it's expected. You'd have to also call the method with concrete values to get the proper assembly (with inlining etc).

**Comment by yancyribbens** [comment]
> Oh, that's not devirtualized but it's expected. You'd have to also call the method with concrete values to get the proper assembly (with inlining etc).

Ah oops.  Thanks for the help with this.

I created a temp crate which "calls"  the rust-bitcoin Amount arbitrary method.  So as not to spam rust-bitcoin, we can offload the conversation here: https://github.com/yancyribbens/arbitrary-bitcoin/issues/1

**Comment by yancyribbens** [comment]
> Yes. I have been meaning to move away from honggfuzz to libfuzzer for this reason but it's not a super high priority.

I noticed rustc uses libfuzzer.  Interestingly, libfuzzer uses Arbitrary internally: https://github.com/rust-fuzz/libfuzzer/blob/main/Cargo.toml#L12.  Given this, I wonder if it would make sense to use Arbitrary with honggfuzz/libfuzzer and proptest::arbitrary with proptest.  We could use derive for both and then use either proptest::arbitrary or arbitrary depending on if cfg=test or not cfg=test.

**Comment by apoelstra** [comment]
I tried playing with libfuzzer, and while I like it much much better as a fuzzer (it has coverage output and minimization, and also I believe it's better at fuzzing, and also it's used by Bitcoin Core so I can reuse a bunch of knowledge), being nightly-only is a tough pill to swallow. I'm unsure how that will interact with my local CI setup, which wants to build every crate in the lockfile using the MSRV compiler.

I did notice that it pulls in an `arbitrary` dep, so if we did use it then it would save us a dependency to just use `arbitrary::Arbitrary` for everything. **But** I don't think this has a ton of value: fuzzer dependencies are not dependencies that downstream users need to deal with (and they have a somewhat limited attack surface even for us as developers).

So I think we'd rather pick one, to reduce the size of our feature matrix and the size of our API. And since the proptest one is strictly better we should stick with that.

**Comment by yancyribbens** [comment]
> I tried playing with libfuzzer, and while I like it much much better as a fuzzer

I prefer libfuzzer too after implementing each.

> So I think we'd rather pick one, to reduce the size of our feature matrix and the size of our API. And since the proptest one is strictly better we should stick with that.

Ok, let's give it a shot!

**Comment by yancyribbens** [comment]
It does not look like we can drop Arbitrary::proptest in place of Arbitrary for honggfuzz because the macro expects an interface to pass unstructured data: https://github.com/rust-fuzz/honggfuzz-rs/blob/master/src/lib.rs#L327

Note in particular this line that takes in a buffer from the fuzzer:
```rust
if let Ok(buf) = Arbitrary::arbitrary(&mut buf) {
```

Maybe I'm too stupid to find it, but I can't find the corresponding interface to pass unstructured data to in proptest::Arbittrary, because if we could, I suppose we could hack together a macro to glue these two incompatible pieces together.

**Comment by yancyribbens** [comment]
Maybe `fill_bytes` from `RngCore` in proptest will work

**Comment by Kixunil** [comment]
Do we really need to figure out fuzzing immediately? We can just use normal `proptest` now and do the rest later.

**Comment by yancyribbens** [comment]
> Do we really need to figure out fuzzing immediately? We can just use normal proptest now and do the rest later.

Sure, I'm fine with that.  The reason I had started working on this was to use it with fuzz testing, although I think proptesting is more useful.

**Comment by yancyribbens** [comment]
Sorry to keep "kicking the tires" on this, but I can't see any reason to not prefer https://github.com/matklad/arbtest.  @apoelstra had mentioned this a while back in this thread and I like how minimalist it is compared to proptest, and to me simplicity means a lot.

To review, proptest and fuzzers may work together using `PassThrough` in `Rng` but the macros don't play well together, so if we use proptest we just leave off using it with a fuzzer until later (and temporary solutions often end up being permanent).  Proptest derive also doesn't play well with fuzzers since it only works in cfg=test while arbtest doesn't have this problem.   Arbtest supports minimization and Arbtest also uses vanilla Arbitrary which would allow us to combine it with whatever fuzzer implementation.  I also note that the original developers of Proptest are largely not involved anymore and there was a long period where the project was abandoned without maintainers.

On the plus side for proptest, people have advocated for proptest because it's version of Arbitrary is seen as strictly better.  Although I'm not sure that it actually gets us anything that we can't do with Arbtest.

Anyway, I'm trying to save a bunch of work in the future by considering the options before opening PR's and creating a bunch of annoying code churn.

**Comment by apoelstra** [comment]
If we are going to implement an `Arbitrary` trait and expose it publicly then it needs to be one of the popular ones. This means either `arbitrary` (from the rust-fuzz org) or the one from `proptest` (by far the most popular property-testing library).

If we just wanted to do internal property testing then sure, we could use an obscure crate and optimize for minimizing disruption.

The complexity of `derive` doesn't matter. We should not be using `derive` except maybe in very simple cases where it provides little value anyway. It slows down compilation, hides logic, and won't even do the right thing because the semantics of Bitcoin objects are not implied by their structural types.

Don't worry about the interaction between the proptest trait and the fuzzer. I'll deal with this. But FYI there is no sense in which property testing is "more useful" or "faster" or anything relative to fuzztesting, or vice-versa.

**Comment by yancyribbens** [comment]
> But FYI there is no sense in which property testing is "more useful" or "faster" or anything relative to fuzztesting, or vice-versa.

Fair point.  I guess imo I feel like in the parlance of coin-selection which is why I went down this rabbit hole with Arbitrary, I feel like property testing is of more value, although this wasn't meant as a broad statement which is how I made it sound.

**Comment by yancyribbens** [comment]
> Gotcha, I had assumed the same but just wanted to double check. I'll read up on it and put something together

@shinghim I don't know if you\re still working on this, but you might be interested in: https://github.com/rust-bitcoin/rust-bitcoin/pull/3015.  We recently added Arbitrary which can be used with arbtest to add a proptest now for Amount formatting.

**Comment by shinghim** [comment]
awesome, nice work on getting the Arbitrary stuff sorted out. I think I'll be able to take a look at this sometime this weekend, but if this is an urgent issue, feel free to work on it

**Comment by yancyribbens** [comment]
~One thing to note is that arbtest requires rustc 1.65. Although I don't think that's a deal breaker since things like fuzz tests used a newer version than the MSRV.~

I was mistaken about this.  In my experiments with Arbtest, there was a different dep that was breaking the MSRV, not arbtest.

**Comment by shinghim** [comment]
Taking a look at the proposed test from [this comment](https://github.com/rust-bitcoin/rust-bitcoin/pull/2604#issuecomment-2204000547):
> I think we should have a proptest checking roughly that `format!("{:$params}", amount.display_in(denom)) == format!("{:$params}", amount.to_sat() as f64 * 10f64.pow(denom.precision()))` for any `params`, `amount` and `denom`

but running into some issues with `f64` precision. Am I correct in assuming that we don't want to bring in another crate like [rust-decimal](https://crates.io/crates/rust_decimal) to deal with floating point precision?

**Comment by apoelstra** [comment]
Correct. And we shouldn't need to, since there are only 51 bits of realistic `Amount` range (21MM times 100MM satoshis) and an IEEE f64 has 52 bits of mantissa.

Arguably if we are having precision issues then this is evidence that our code is wrong.

**Comment by yancyribbens** [comment]
> Am I correct in assuming that we don't want to bring in another crate like [rust-decimal](https://crates.io/crates/rust_decimal) to deal with floating point precision?

For the testing framework, we've included things like Arbitrary that we wouldn't normally include.  You could try including rust-decimal just to figure out how the code should work then try to do without it.

**Comment by shinghim** [comment]
> Arguably if we are having precision issues then this is evidence that our code is wrong.

the issue I was having is occurring when converting the sats from `u64` to `f64` to verify our code: `amount.to_sat() as f64 * 10f64.pow(denom.precision())`. For example, when displaying `560937.01792084` as bits, we display `560937017920.8400000`, but we are getting some slightly imprecise floating-point values to verify against (`560937017920.8398438`) which causes it to fail. AFAICT our code is working fine, just need to figure out a way around the floating point precision

**Comment by tcharding** [comment]
I wonder if we should have an `Amount` type in `units` that is used in `primitives` (implies in all primitive data structures) which uses sats only, has no float stuff at all (no `from_btc` constructors). Then we have a `DasplayableAmount` (with better name) that provides parsing and displaying.

We could possibly then further separate the amount type that is "this is a realistic single bitcoin amount" ie. < 21,000,000 and the "this is for summing multiple amounts" type that can be any value.

This would result in a super simple `Amount` type in `units`.

**Comment by Kixunil** [comment]
If we should then both types still belong to `units` so we don't get any simplification benefits. I think we should stick to the current design at this point. We can declare that our `Display` impl is unstable for now and the only thing we promise is that unless you mess with `precision` we will display an accurate human-readable number and that the serialization roundtrips with `FromStr`..

**Comment by apoelstra** [comment]
Agreed with @Kixunil on all counts.

**Comment by tcharding** [comment]
Sounds reasonable, thanks.
